# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [1]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf
import statsmodels.api as sm

In [2]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [3]:
reg = smf.ols('tip ~ net_bill', data = tips).fit()
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    tip   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.330
Method:                 Least Squares   F-statistic:                     120.6
Date:                Mon, 22 Jan 2024   Prob (F-statistic):           5.02e-23
Time:                        19:29:23   Log-Likelihood:                -375.63
No. Observations:                 244   AIC:                             755.3
Df Residuals:                     242   BIC:                             762.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.3307      0.168      7.907      0.000       0.999       1.662
net_bill       0.0993      0.009     10.980      0.000       0.082       0.117
==============================================================================
Omnibus:                       33.873   Durbin-Watson:                   2.123
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               59.942
Skew:                           0.754   Prob(JB):                     9.63e-14
Kurtosis:                       4.903   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
# Adicionando as colunas tip_pct e net_bill
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

In [7]:
# Ajustando o modelo de regressão de tip_pct por net_bill
X = sm.add_constant(tips['net_bill'])
y_tip_pct = tips['tip_pct']
model_tip_pct = sm.OLS(y_tip_pct, X).fit()

In [8]:
# Obtendo os valores preditos para tip_pct
tips['pred_tip_pct'] = model_tip_pct.fittedvalues

In [9]:
# Obtendo os valores preditos para tip
tips['pred_tip'] = tips['pred_tip_pct'] * tips['net_bill']

In [10]:
# Calculando o R² para o modelo de tip_pct
r_squared_tip_pct = model_tip_pct.rsquared
print(f"R² para tip_pct: {r_squared_tip_pct}")

R² para tip_pct: 0.09129084920662933


In [11]:
# Calculando o R² para o modelo de tip
X_tip = sm.add_constant(tips['pred_tip_pct'])
y_tip = tips['tip']
model_tip = sm.OLS(y_tip, X_tip).fit()
r_squared_tip = model_tip.rsquared
print(f"R² para tip: {r_squared_tip}")

R² para tip: 0.3325407312323705


### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [12]:
# Ajustando o modelo de regressão de tip_pct por log(net_bill)
tips['log_net_bill'] = np.log(tips['net_bill'])
X_log = sm.add_constant(tips['log_net_bill'])
model_tip_pct_log = sm.OLS(y_tip_pct, X_log).fit()

In [13]:
# Obtendo os valores preditos para tip_pct usando log(net_bill)
tips['pred_tip_pct_log'] = model_tip_pct_log.fittedvalues

In [14]:
# Calculando o R² para o modelo de tip_pct usando log(net_bill)
r_squared_tip_pct_log = model_tip_pct_log.rsquared
print(f"R² para tip_pct usando log(net_bill): {r_squared_tip_pct_log}")

R² para tip_pct usando log(net_bill): 0.20172884237605493


###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [15]:
# Ajustando o modelo de regressão de tip_pct por um polinômio de segundo grau em net_bill
tips['net_bill_squared'] = tips['net_bill'] ** 2
X_poly = sm.add_constant(tips[['net_bill', 'net_bill_squared']])
model_tip_pct_poly = sm.OLS(y_tip_pct, X_poly).fit()

In [16]:
# Obtendo os valores preditos para tip_pct usando o polinômio de segundo grau
tips['pred_tip_pct_poly'] = model_tip_pct_poly.fittedvalues

In [17]:
# Calculando o R² para o modelo de tip_pct usando o polinômio de segundo grau
r_squared_tip_pct_poly = model_tip_pct_poly.rsquared
print(f"R² para tip_pct usando polinômio de segundo grau em net_bill: {r_squared_tip_pct_poly}")

R² para tip_pct usando polinômio de segundo grau em net_bill: 0.15760784734296185


###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [18]:
# Ajustando o modelo de regressão do logaritmo de tip_pct por net_bill
model_log_tip_pct = sm.OLS(np.log(tips['tip_pct']), sm.add_constant(tips['net_bill'])).fit()

In [19]:
# Obtendo os valores preditos para o logaritmo de tip_pct
tips['pred_log_tip_pct'] = model_log_tip_pct.fittedvalues

In [20]:
# Calculando o R² para o modelo do logaritmo de tip_pct por net_bill
r_squared_log_tip_pct = model_log_tip_pct.rsquared
print(f"R² para o logaritmo de tip_pct por net_bill: {r_squared_log_tip_pct}")

R² para o logaritmo de tip_pct por net_bill: 0.23534487140864324


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [21]:
# Ajustando o modelo de regressão do logaritmo de tip_pct por logaritmo de net_bill
model_log_log_tip_pct = sm.OLS(np.log(tips['tip_pct']), sm.add_constant(np.log(tips['net_bill']))).fit()

In [22]:
# Obtendo os valores preditos para o logaritmo de tip_pct
tips['pred_log_log_tip_pct'] = model_log_log_tip_pct.fittedvalues

In [23]:
# Calculando o R² para o modelo do logaritmo de tip_pct por logaritmo de net_bill
r_squared_log_log_tip_pct = model_log_log_tip_pct.rsquared
print(f"R² para o logaritmo de tip_pct por logaritmo de net_bill: {r_squared_log_log_tip_pct}")

R² para o logaritmo de tip_pct por logaritmo de net_bill: 0.2934740215080778


###  6. qual modelo resultou no melhor ajuste?

No meu caso, o melhor modelo foi o ponto 5 com R**2 0.29.